atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CqlClientImpl.java

```diff
@@ -0,0 +1,336 @@
+    private final PreparedStatement healthCheckStatement;
+    private final Cluster cluster;
+    private final Session session;
+
+    public static CqlClient create(CassandraKeyValueServiceConfig config, Set<InetSocketAddress> servers) {
+        long curId = cassandraId.getAndIncrement();
+        String clusterName = "cassandra" + "-" + curId;
+        CqlClientImpl.SimpleAddressTranslator mapper = new CqlClientImpl.SimpleAddressTranslator(config);
+
+        Cluster.Builder clusterBuilder = Cluster.builder()
+                .addContactPointsWithPorts(servers)
+                .withClusterName(clusterName)
+                .withCredentials(config.credentials().username(), config.credentials().password())
+                .withCompression(ProtocolOptions.Compression.LZ4)
+                .withLoadBalancingPolicy(loadBalancingPolicy(config, servers))
+                .withPoolingOptions(poolingOptions(config))
+                .withQueryOptions(queryOptions(config))
+                .withRetryPolicy(DefaultRetryPolicy.INSTANCE)
+                .withAddressTranslator(mapper)
+                .withoutJMXReporting()
+                .withThreadingOptions(new ThreadingOptions());
+
+        sslOptions(config).ifPresent(clusterBuilder::withSSL);
+
+        Cluster cluster = buildCluster(clusterBuilder);
+
+        String sessionName = cluster.getClusterName()
+                + "-session";
+
+        Session session = cluster.connect();
+
+        return create(sessionName, cluster, session);
+    }
+
+    public static CqlClient create(String clientName, Cluster cluster, Session session) {
+        ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                .setNameFormat(clientName + "-%d")
+                .build();
+
+        return create(clientName, cluster, session, threadFactory);
+    }
+
+    public static CqlClient create(String clientName, Cluster cluster, Session session,
+            ThreadFactory threadFactory) {
+        return create(clientName, cluster, session, Executors.newCachedThreadPool(threadFactory));
+    }
+
+    public static CqlClient create(String clientName, Cluster cluster, Session session,
+            Executor executor) {
+        PreparedStatement healthCheckStatement = session.prepare("SELECT dateof(now()) FROM system.local ;");
+        return new CqlClientImpl(clientName, cluster, session, executor, healthCheckStatement).start();
+    }
+
+    private CqlClientImpl(String clientName, Cluster cluster, Session session, Executor executor,
+            PreparedStatement healthCheckStatement) {
+        this.clientName = clientName;
+        this.cluster = cluster;
+        this.session = session;
+        this.executor = executor;
+        this.healthCheckStatement = healthCheckStatement;
+    }
+
+
+    @Override
+    public String sessionName() {
+        return clientName;
+    }
+
+    @Override
+    public void close() {
+        healthCheckExecutor.shutdownNow();
```

Comment:
so the close method here *only* deals with the health check, when actually, we also need to deal with the underlying `Session` that should also be closed.